error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoZipOfMonoVoidUsage.java

```diff
@@ -0,0 +1,130 @@
+package tech.picnic.errorprone.bugpatterns;
+
+import static com.google.errorprone.BugPattern.LinkType.CUSTOM;
+import static com.google.errorprone.BugPattern.SeverityLevel.ERROR;
+import static com.google.errorprone.BugPattern.StandardTags.LIKELY_ERROR;
+import static com.google.errorprone.matchers.Matchers.allOf;
+import static com.google.errorprone.matchers.Matchers.anyMethod;
+import static com.google.errorprone.matchers.Matchers.staticMethod;
+import static com.google.errorprone.matchers.Matchers.toType;
+import static java.util.stream.Collectors.joining;
+import static tech.picnic.errorprone.bugpatterns.util.Documentation.BUG_PATTERNS_BASE_URL;
+import static tech.picnic.errorprone.bugpatterns.util.MoreTypes.generic;
+import static tech.picnic.errorprone.bugpatterns.util.MoreTypes.type;
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.MemberReferenceTreeMatcher;
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.suppliers.Supplier;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MemberReferenceTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.tools.javac.code.Type;
+import java.util.stream.Stream;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+
+/**
+ * A {@link BugChecker} that flags usages of {@link Mono#zip(Mono, Mono)}} and {@link
+ * Mono#zipWith(Mono)}} with {@link Mono#empty()} parameters.
+ *
+ * <p>{@link Mono#zip(Mono, Mono)} and {@link Mono#zipWith(Mono)} perform incorrectly upon retrieval
+ * of the empty publisher and prematurely terminates the reactive chain from the execution. In most
+ * cases this is not the desired behaviour and {@link Mono#concatWith(Publisher)} or {@link
+ * Mono#then(Mono)} should be preferred, as it produces consistent results and performs in the
+ * predictable manner.
+ *
+ * <p>NB: Mono&lt;?>#zipWith(Mono&lt;Void>) is allowed be the Reactor API, but it is an incorrect
+ * usage of the API. It will be replaced with Mono&lt;?>#concatWith(Mono&lt;Void>) but it will lead
+ * to the compilation errors, indicating that the logic under the hood should be revisited and
+ * fixed.
```

Comment:
Discussed with @DimaLegeza offline. Flagging cases of which we can be quite sure that it indicates a problem sound like a nice plan.  

We were wondering if you have a solution / idea about this problem https://github.com/PicnicSupermarket/error-prone-support/pull/733#discussion_r1277458027 @Stephan202. 

@DimaLegeza will play around with trying to get the logic for flagging the likely error cases right. After that he'll push the changes for a proposal for the implementation of this check.

Reply:
Nice to see this PR resurrected!

> We were wondering if you have a solution / idea about this problem [#733 (comment)](https://github.com/PicnicSupermarket/error-prone-support/pull/733#discussion_r1277458027) @Stephan202.

@oxkitsune's initial suggestion there is correct. The fact that it's [not enough](https://github.com/PicnicSupermarket/error-prone-support/pull/733#discussion_r1277498380), shows that type matching won't suffice: one would need to look at the "shape" of the expressions, e.g. by explicitly looking for `Mono.empty()`.

But... since this PR was opened we introduced the [`IsEmpty`](https://github.com/PicnicSupermarket/error-prone-support/blob/1d25b78c4c6bbc426b32766aaa373c1361a45a6e/refaster-support/src/main/java/tech/picnic/errorprone/refaster/matchers/IsEmpty.java) matcher, which already [recognizes](https://github.com/PicnicSupermarket/error-prone-support/blob/1d25b78c4c6bbc426b32766aaa373c1361a45a6e/refaster-support/src/main/java/tech/picnic/errorprone/refaster/matchers/IsEmpty.java#L98-L104) `Mono.isEmpty()`, and it has a [TODO](https://github.com/PicnicSupermarket/error-prone-support/blob/1d25b78c4c6bbc426b32766aaa373c1361a45a6e/refaster-support/src/main/java/tech/picnic/errorprone/refaster/matchers/IsEmpty.java#L52-L53) for supporting e.g. `Mono<Void>`. So I see two ways forward:
1. Just write a few Refaster rules for the most common cases (extending `IsEmpty` if necessary).
2. Use the `IsEmpty` logic as a foundation for a `BugChecker`.